store/sqlstore/channel_store.go

```diff
@@ -3950,46 +3948,175 @@ func (s SqlChannelStore) UpdateSidebarCategories(userId, teamId string, categori
+		return nil
+	}
+
+	// Delete any corresponding SidebarChannels entries in a Favorites category corresponding to this preference. This
+	// can't use the query builder because it uses DB-specific syntax
+	params := map[string]interface{}{
+		"UserId":       preference.UserId,
+		"ChannelId":    preference.Name,
+		"CategoryType": model.SidebarCategoryFavorites,
+	}
+	var query string
+	if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		query = `
+			DELETE
+				SidebarChannels
+			FROM
+				SidebarChannels
+			JOIN
+				SidebarCategories ON SidebarChannels.CategoryId = SidebarCategories.Id
+			WHERE
+				UserId = :UserId
+				AND ChannelId = :ChannelId
+				AND Type = :CategoryType`
+	} else {
+		query = `
+			DELETE FROM
+				SidebarChannels
+			USING
+				SidebarCategories
+			WHERE
+				SidebarChannels.CategoryId = SidebarCategories.Id
+				AND SidebarChannels.UserId = :UserId
+				AND SidebarChannels.ChannelId = :ChannelId
+				AND SidebarCategories.Type = :CategoryType`
+	}
+
+	if _, err := transaction.Exec(query, params); err != nil {
+		return errors.Wrap(err, "Failed to remove sidebar entries for preference")
+	}
+
+	return nil
+}
+
+func (s SqlChannelStore) addChannelToFavoritesCategory(transaction *gorp.Transaction, preference *model.Preference) error {
+	if preference.Category != model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL {
+		return nil
+	}
+
+	var channel *model.Channel
+	if obj, err := transaction.Get(&model.Channel{}, preference.Name); err != nil {
+		return errors.Wrapf(err, "Failed to get favorited channel with id=%s", preference.Name)
+	} else {
+		channel = obj.(*model.Channel)
+	}
+
+	var teamIdFilter sq.Sqlizer
+	if channel.TeamId == "" {
+		// Accept the Favorites category on any team
+		teamIdFilter = sq.Expr("(1=1)")
+	} else {
+		teamIdFilter = sq.Eq{"TeamId": channel.TeamId}
+	}
+
+	// Get the IDs of the Favorites category/categories that the channel needs to be added to
+	idsQuery, idsParams, _ := s.getQueryBuilder().
+		Select("Id").
+		From("SidebarCategories").
+		Where(sq.And{
+			teamIdFilter,
+			sq.Eq{
+				"UserId": preference.UserId,
+				"Type":   model.SidebarCategoryFavorites,
+			},
+			sq.Expr("Id NOT IN (select CategoryId from SidebarChannels where ChannelId = ?)", channel.Id),
+		}).ToSql()
+
+	var categoryIds []string
+	if _, err := transaction.Select(&categoryIds, idsQuery, idsParams...); err != nil {
+		return errors.Wrap(err, "Failed to get Favorites sidebar categories")
+	}
+
+	if len(categoryIds) == 0 {
+		// The channel is already in the Favorites category/categories
+		return nil
+	}
+
+	// For each category ID, insert a row into SidebarChannels with the given channel ID and a SortOrder that's less than
+	// all existing SortOrders in the category so that the newly favorited channel comes first
+	query, params, _ := s.getQueryBuilder().
+		Insert("SidebarChannels").
+		Columns(
+			"ChannelId",
+			"CategoryId",
+			"UserId",
+			"SortOrder",
```

Comment:
Stepping back quite a bit here, but I'm curious if we need an explicit `SidebarChannels` table with a categorized channel per row. My theory is that we almost always bulk-query this table by `UserId`, and rarely perform row-level access.

Should we consider a `SidebarCategories.Channels` JSON column that lists the `Ids` instead? This would dramaticlly simplify the `SortOrder` juggling, since the order would be implicit in the JSON array. 

I suspect it would also dramatically improve performance. I'm worried about `SidebarChannels` as a table, since I'm assuming this will approach the size of the existing `Channels` table when the feature is heavily used, and we've had performance issues there.

Reply:
As mentioned in a DM, this isn't going to be something we can fix in 5.26 at this point. Since this will only affect people with the feature enabled, it shouldn't have any major negative impacts in the short term. At the very least, the client is unaware of any of this, so we're free to change up the database implementation in the future as necessary.